stdlib/src/test/scala/chisel3/std/axi4/TestMem.scala

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.std.axi4
+
+import chisel3._
+import chisel3.stage.ChiselStage
+import chisel3.util._
+import org.scalatest.flatspec.AnyFlatSpec
+
+import scala.annotation.tailrec
+
+/** Simple test to make sure that we can properly instantiate and elaborate the Axi4 bundles. */
+class Axi4MemTests extends AnyFlatSpec {
+  behavior.of("Axi4TestMem")
+
+  it should "elaborate" in {
+    val params = Axi4IOParameters(addressBits = 32, dataBits = 32, idBits = 0)
+    val targetDir = "test_run_dir/Axi4TestMem_should_elaborate"
+    val args = Array("-td", targetDir, "--full-stacktrace")
+    (new ChiselStage).emitSystemVerilog(new Axi4TestMem(params, baseAddress = 0x1000, mask = 0xff), args)
+  }
+}
+
+/** A simple Axi4 memory based on the Axi4 SRAM from rocket-chip. */
+class Axi4TestMem(params: Axi4IOParameters, baseAddress: BigInt, mask: BigInt) extends Module {
+  val io = IO(Flipped(new Axi4IO(params)))
+
+  // TODO: allow 0-bit signals to be unconnected
+  io.readData.bits.user := DontCare
+  io.writeResponse.bits.user := DontCare
+
+  // parameter calculations
+  val beatBytes = params.dataBits / 8
+  @tailrec
+  private def toBitList(x: BigInt, tail: List[Boolean] = Nil): List[Boolean] =
+    if (x == 0) tail.reverse else toBitList(x >> 1, ((x & 1) == 1) :: tail)
+  val maskBits = toBitList(mask >> log2Ceil(beatBytes))
+  val memoryEntries = BigInt(1) << maskBits.count(b => b)
+
+  val mem = SyncReadMem(memoryEntries, Vec(beatBytes, UInt(8.W)))
+
+  private def decodeAddress(address: UInt): (UInt, Bool) = {
+    val a = Cat(maskBits.zip((address >> log2Ceil(beatBytes)).asBools).filter(_._1).map(_._2).reverse)
+    val sel = ((address ^ baseAddress.U).zext & (~mask).S) === 0.S
+    (a, sel)
+  }
+  val (r_addr, r_sel0) = decodeAddress(io.readAddress.bits.address)
+  val (w_addr, w_sel0) = decodeAddress(io.writeAddress.bits.address)
```

Comment:
#2449 merged, I think we can have a try to `BitSet. fromRange` and `decoder.bitset` to provide a better PPA.

Reply:
Yes, there are probably a lot of possible optimizations. Note though, that the aim of this module is only to serve as a testing ground for the bundle definitions. If we never use the bundle to generate HW, it will not be tested properly.

Reply:
Sure!

Reply:
If we ever decide to actually offer the IPs in the library (and just in `test`) we will have to be concerned more about PPA.
But also, since this would be the standard library we would have to make sure not to optimize for one particular backend. Things should synthesize reasonably well for FPGA and various ASIC technologies.